pallets/asset-index/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2021 ChainSafe Systems
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub mod traits;
+mod types;
+
+#[frame_support::pallet]
+// this is requires as the #[pallet::event] proc macro generates code that violates this lint
+#[allow(clippy::unused_unit)]
+pub mod pallet {
+    use crate::traits::AssetRecorder;
+    use crate::types::{AssetAvailability, IndexAssetData, PendingRedemption};
+    use frame_support::{
+        dispatch::DispatchResultWithPostInfo,
+        pallet_prelude::*,
+        sp_runtime::traits::{CheckedAdd, Zero},
+        traits::{Currency, LockableCurrency},
+    };
+    use frame_system::pallet_prelude::*;
+
+    type AccountIdFor<T> = <T as frame_system::Config>::AccountId;
+    type BalanceFor<T> = <<T as Config>::IndexToken as Currency<AccountIdFor<T>>>::Balance;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        // Origin that is allowed to administer the index
+        type AdminOrigin: EnsureOrigin<Self::Origin>;
+        // Currency implementation to use as the index token
+        type IndexToken: LockableCurrency<Self::AccountId>;
+        // Period after the minting of the index token for which 100% is locked up.
+        // Only applies to users contributing assets directly to index
+        type LockupPeriod: Get<Self::BlockNumber>;
+        // The minimum amount of the index token that can be redeemed for the underlying asset in the index
+        type MinimumRedemption: Get<BalanceFor<Self>>;
+        // Minimum amount of time between redeeming index tokens
+        // and being able to withdraw the awarded assets
+        type WithdrawalPeriod: Get<Self::BlockNumber>;
+        // The maximum amount of DOT that can exist in the index
+        type DOTContributionLimit: Get<BalanceFor<Self>>;
+        // Type used to identify assets
+        type AssetId: Parameter + Encode + Decode;
```

Comment:
I believe `Parameter` is already `: Codec`, but maybe should also require `Member`?
```suggestion
        type AssetId: Parameter;
```